beam/src/main/scala/magnolify/beam/BeamSchemaType.scala

```diff
@@ -0,0 +1,276 @@
+    protected def buildSchema(cm: CaseMapper): Schema
+    def schema(cm: CaseMapper): Schema = schemaCache.getOrElseUpdate(cm.uuid, buildSchema(cm))
+  }
+
+  // ////////////////////////////////////////////////
+
+  type Typeclass[T] = BeamSchemaField[T]
+  implicit def gen[T]: BeamSchemaField[T] = macro Magnolia.gen[T]
+
+  // TODO beam schemas support OneOf
+  @implicitNotFound("Cannot derive BeamSchemaField for sealed trait")
+  private sealed trait Dispatchable[T]
+  def split[T: Dispatchable](sealedTrait: SealedTrait[Typeclass, T]): BeamSchemaField[T] = ???
+  //  new BeamSchemaField[T] {
+  //    override type FromT = ???
+  //    override type ToT = ???
+  //    override def fieldType(cm: CaseMapper): FieldType = {
+  //      FieldType.logicalType(
+  //        logicaltypes.OneOfType.create(
+  //          sealedTrait.subtypes.map { sub =>
+  //            Field.of(s"${sub.typeName.owner}.${sub.typeName.short}", sub.typeclass.fieldType(cm))
+  //          }
+  //          .asJava
+  //        )
+  //      )
+  //    }
+  //    override def from(v: this.type)(cm: CaseMapper): T = ???
+  //    override def to(v: T)(cm: CaseMapper): this.type = ???
+  //  }
+
+  def join[T](caseClass: CaseClass[Typeclass, T]): BeamSchemaField[T] = {
+    if (caseClass.isValueClass) {
+      val p = caseClass.parameters.head
+      val tc = p.typeclass
+      new BeamSchemaField[T] {
+        override type FromT = tc.FromT
+        override type ToT = tc.ToT
+        override def fieldType(cm: CaseMapper): FieldType = tc.fieldType(cm)
+        override def from(v: FromT)(cm: CaseMapper): T = caseClass.construct(_ => tc.fromAny(v)(cm))
+        override def to(v: T)(cm: CaseMapper): ToT = tc.to(p.dereference(v))(cm)
+      }
+    } else {
+      new Record[T] {
+        override def fieldType(cm: CaseMapper): FieldType = FieldType.row(schema(cm))
+
+        override protected def buildSchema(cm: CaseMapper): Schema =
+          caseClass.parameters
+            .foldLeft(Schema.builder()) { case (s, p) =>
+              s.addField(cm.map(p.label), p.typeclass.fieldType(cm))
+            }
+            .build()
+
+        override def from(v: Row)(cm: CaseMapper): T =
+          caseClass.construct(p => p.typeclass.fromAny(v.getValue[Any](p.index))(cm))
+
+        override def to(v: T)(cm: CaseMapper): Row = {
+          val zero: Either[Row.Builder, Row.FieldValueBuilder] = Left(Row.withSchema(schema(cm)))
+          val eitherBuilder = caseClass.parameters
+            .foldLeft(zero) { (eitherBuilder, p) =>
+              val value = p.typeclass.to(p.dereference(v))(cm)
+              eitherBuilder match {
+                case Left(rowBuilder)    => Right(rowBuilder.withFieldValue(p.index, value))
+                case Right(fieldBuilder) => Right(fieldBuilder.withFieldValue(p.index, value))
+              }
+            }
+          eitherBuilder match {
+            case Left(rb)  => rb.build()
+            case Right(fb) => fb.build()
+          }
```

Comment:
```suggestion
          val values = caseClass.parameters.map { p =>
            p.typeclass.to(p.dereference(v))(cm)
          }
          Row.withSchema(schema(cm)).addValues(values: _*).build()
```

Reply:
Much better and obvious in retrospect, thanks.